### ISSUE TITLE
Propagate line and column number to the type checker

### ISSUE THREAD

**Issue by zhiyuanshi** [hkuplg/fcore#90]
Currently line and column number is attached to each token, but discarded in the parser during the construction of abstract syntax trees.

**Comment by zonyitoo** [comment]
+1

**Comment by zhiyuanshi** [comment]
@wxzh Do you have time to take a look at this?

**Comment by zhiyuanshi** [comment]
I think we should raise the urgency of this issue. Otherwise it'll be hard to work in our language.

**Comment by bixuanzju** [comment]
+1

**Comment by brunocdsoliveira** [comment]
Your are the issue manager :). If you think it is urgent, make it urgent!

One thing that may help on the meantime is to also print the expression with the type-error. Although this is not as precise as a line number, it will definitely help pinpointing where the problem is (and I guess it is easier to implement).

**Comment by zhiyuanshi** [comment]
@brunocdsoliveira copy :+1:

**Comment by wxzh** [comment]
Partially done. For now, most cases are covered except those without expression or location information in type-checking. 

Possible further improvement: 
1. Tagging types with location
2. Tagging names with location
3. Change type-checking monad to trace current expression
4. Improve the accuracy of the location

**Comment by zonyitoo** [comment]
@wxzh There still has many situation that the compiler won't report line:column when type mismatch.

It's hard for me to make a minimal test case for you, but what I just do is to modify the type parameters of a Algebraic Data Type. I count on the compiler to find those mismatching on functions for me, but seems it doesn't work.

**Comment by zonyitoo** [comment]
Why I unassigned @wxzh ? I haven't touched anything!

One situation, say I have a function

```
let fn[A] (a : A) = ...
```

and I also have a type/data

```
type NewType[B] = Const
```

When I call the `fn` by `fn[NewType]  xxx`, then compiler may reports something like this

```
error: kind mismatch:
  expected ‘*’,
     found ‘(* => *)’
```

without line:column

**Comment by wxzh** [comment]
@zonyitoo This is because we do not trace the location of type for now as listed in possible improvement 1. It is possible to give a more precise location information about type but this will influence a lot of existed code. I am not sure whether to modify or not.

For kind mismatch, I think reporting the type name that causes the error can help a little bit. I'll change it.

**Comment by zonyitoo** [comment]
@wxzh alright. So I cannot count on the compiler to help me find out which one should be modified. Ah ...

**Comment by zonyitoo** [comment]
Also, there still have some situations of type mismatch that the compiler won't report the location of. I cannot reproduce it now because that would cause a huge mess of my code (over 1000 lines).

**Comment by wxzh** [comment]
@zonyitoo The only thing the compiler can tell you is that the two types' kinds are different. _You_ should decide which one should be modified.

**Comment by wxzh** [comment]
@zonyitoo Check whether b95a307 provides the information you want

**Comment by zonyitoo** [comment]
@wxzh Seems nothing changed, still report kind mismatch without location.

**Comment by wxzh** [comment]
@zonyitoo I know. What I added for _kind_ mismatch is printing the ill-kinded type. For now, all _type_ mismatches should have location and ill-typed expression.

**Comment by zonyitoo** [comment]
@wxzh I still get this output

```
error: kind mismatch:
  expected ‘*’,
     found ‘(* => *)’
```

**Comment by wxzh** [comment]
@zonyitoo My pull request has not been merged yet ..

**Comment by bixuanzju** [comment]
@zonyitoo It's merged, have a try!

**Comment by zonyitoo** [comment]
@wxzh @bixuanzju Helpful for small program, but when I try to delete one kind param in 1000 line, then I won't be able to find it, even I know what exactly the kind mismatch is...

**Comment by bixuanzju** [comment]
Do you think it can be closed?

**Comment by zhiyuanshi** [comment]
Nope. Still, some nodes in the AST are not annotated with location. The task is two-fold:
1. Annotate all AST nodes with location;
2. Report location in all error messages.
